browser/permissions/localhost_access_permission_browsertest.cc

```diff
@@ -0,0 +1,287 @@
+  void CheckNoPromptFlow(std::string button_id, GURL localhost_url) {
+    CheckCurrentStatusIs(ContentSetting::CONTENT_SETTING_ASK);
+    EXPECT_EQ(0, prompt_factory()->show_count());
+    ASSERT_TRUE(ui_test_utils::NavigateToURL(browser(), embedding_url_));
+    // No prompt.
+    prompt_factory()->set_response_type(
+        permissions::PermissionRequestManager::DENY_ALL);
+    ClickButtonWithId(button_id);
+    // Make sure prompt did not come up.
+    EXPECT_EQ(0, prompt_factory()->show_count());
+    // Check content setting is still ASK.
+    CheckCurrentStatusIs(ContentSetting::CONTENT_SETTING_ASK);
+  }
+
+ protected:
+  GURL embedding_url_;
+  content::ContentMockCertVerifier mock_cert_verifier_;
+  std::unique_ptr<net::EmbeddedTestServer> https_server_;
+  base::test::ScopedFeatureList feature_list_;
+  Browser* current_browser_;
+
+ private:
+  std::unique_ptr<permissions::MockPermissionPromptFactory> prompt_factory_;
+};
+
+std::unique_ptr<net::test_server::HttpResponse> HandleHttpRequest(
+    const std::string& domain,
+    const net::test_server::HttpRequest& request) {
+  const auto port = request.GetURL().port();
+  auto http_response = std::make_unique<net::test_server::BasicHttpResponse>();
+  http_response->set_code(net::HTTP_OK);
+  http_response->set_content_type("text/html");
+  if (request.GetURL().path() == kTestEmbeddingPath) {
+    http_response->set_content(
+        base::StringPrintf(kTestBodyScaffolding, domain.c_str(), port.c_str()));
+  } else if (request.relative_url == "target.html") {
+    // Simple response for localhost.
+    http_response->set_content("This is a localhost page");
+  }
+  return http_response;
+}
+
+IN_PROC_BROWSER_TEST_F(LocalhostAccessBrowserTest, Localhost) {
+  std::string test_domain = "localhost";
+  https_server_->RegisterRequestHandler(
+      base::BindRepeating(HandleHttpRequest, test_domain));
+  ASSERT_TRUE(https_server_->Start());
+  embedding_url_ =
+      https_server_->GetURL(kTestEmbeddingDomain, kTestEmbeddingPath);
+  const auto& target_url = https_server_->GetURL(test_domain, kTestTargetPath);
+  CheckAskAndAcceptFlow(kButtonHtmlId, target_url);
+  // Reset content setting
+  SetCurrentStatus(ContentSetting::CONTENT_SETTING_ASK);
+  CheckAskAndDenyFlow(kButtonHtmlId, target_url, 1);
+}
+
+IN_PROC_BROWSER_TEST_F(LocalhostAccessBrowserTest, DotLocalhost) {
+  std::string test_domain = "test.localhost";
+  https_server_->RegisterRequestHandler(
+      base::BindRepeating(HandleHttpRequest, test_domain));
+  ASSERT_TRUE(https_server_->Start());
+  embedding_url_ =
+      https_server_->GetURL(kTestEmbeddingDomain, kTestEmbeddingPath);
+  const auto& target_url = https_server_->GetURL(test_domain, kTestTargetPath);
+  CheckAskAndAcceptFlow(kButtonHtmlId, target_url);
+  // Reset content setting
+  SetCurrentStatus(ContentSetting::CONTENT_SETTING_ASK);
+  CheckAskAndDenyFlow(kButtonHtmlId, target_url, 1);
+}
+
+IN_PROC_BROWSER_TEST_F(LocalhostAccessBrowserTest, OneTwoSeven) {
```

Comment:
Not a big deal, but for completeness' sake, we could also test the IPv6 version (`[::1]`).